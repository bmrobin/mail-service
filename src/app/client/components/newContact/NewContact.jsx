import React from 'react';
import { Form, FormControl, FormGroup, ControlLabel, Button, Col } from 'react-bootstrap';
import Error from 'components/error/Error';
import { saveUser } from 'utils/api';
import './newContact.css';

export default class NewContact extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      connectionError: false
    };
  }

  saveEmail(event) {
    event.preventDefault();
    saveUser(this.state.emailAddress).then(() => {
      this.setState({ connectionError: false });
      this.resetForm();
    }).catch(() => {
      this.resetForm();
      this.setState({ connectionError: true });
    });
  }

  resetForm() {
    document.getElementById('email-id').value = '';
    document.getElementById('save-id').blur();
  }

  handler(event) {
    this.setState({
      emailAddress: event.target.value,
      connectionError: false
    });
  }

  render() {
    return (
      <Col xs={12} md={7} className="email-form">
        <Form horizontal onSubmit={(e) => this.saveEmail(e)}>

          <FormGroup>
            <Col sm={2} componentClass={ControlLabel}>
              Email
            </Col>
            <Col sm={10}>
              <FormControl type="text" id="email-id" onChange={(e) => this.handler(e)} />
            </Col>
          </FormGroup>

          {this.state.connectionError && <Error message="An error occurred while saving. Check connection to server." />}

          <Button type="submit" className="save-btn" id="save-id">Save</Button>

        </Form>
      </Col>
    );
  }
}
