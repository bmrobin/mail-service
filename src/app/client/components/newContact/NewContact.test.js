import React from 'react';
import { shallow } from 'enzyme';
import { FormControl, Form } from 'react-bootstrap';
import NewContact from './NewContact';

jest.mock('utils/api', () => (
  {
    saveUser: jest.fn((args) => {
      return new Promise((resolve) => {
        resolve();
      });
    })
  }
));

describe("NewContact Component", () => {

  test("should render component", () => {
    const component = shallow(<NewContact />);
    expect(component.debug()).toMatchSnapshot();
  });

  test("should store email address when changed in form", () => {
    const component = shallow(<NewContact />);
    component.find(FormControl).simulate('change', { target: { value: 'bmrobin@mail.com' }});
    expect(component.state('emailAddress')).toBe('bmrobin@mail.com');
    expect(component.state('connectionError')).toBe(false);
  });

  test("should create new contact when form is submitted", () => {
    const component = shallow(<NewContact />);
    const mockEventFn = jest.fn();
    component.setState({ connectionError: true });
    component.find(FormControl).simulate('change', { target: { value: 'bmrobin@mail.com' }});
    component.find(Form).simulate('submit', { preventDefault : mockEventFn });
    expect(mockEventFn).toHaveBeenCalledTimes(1);
    expect(component.state('connectionError')).toBe(false);
  });

});
