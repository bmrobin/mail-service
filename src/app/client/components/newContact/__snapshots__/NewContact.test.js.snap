// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewContact Component should render component 1`] = `
"<Col xs={12} md={7} className=\\"email-form\\" componentClass=\\"div\\" bsClass=\\"col\\">
  <Form horizontal={true} onSubmit={[Function: onSubmit]} inline={false} componentClass=\\"form\\" bsClass=\\"form\\">
    <FormGroup bsClass=\\"form-group\\">
      <Col sm={2} componentClass={[Function: ControlLabel]} bsClass=\\"col\\">
        Email
      </Col>
      <Col sm={10} componentClass=\\"div\\" bsClass=\\"col\\">
        <FormControl type=\\"text\\" id=\\"email-id\\" onChange={[Function: onChange]} componentClass=\\"input\\" bsClass=\\"form-control\\" />
      </Col>
    </FormGroup>
    <Button type=\\"submit\\" className=\\"save-btn\\" id=\\"save-id\\" active={false} block={false} disabled={false} bsStyle=\\"default\\" bsClass=\\"btn\\">
      Save
    </Button>
  </Form>
</Col>"
`;
